src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -1531,9 +1541,71 @@ Collection<CompactionAggregate.UnifiedAggregate> getCompactionAggregates(Arena a
                                                                       this::makeBucket,
                                                                       unbucketed::addAll);
 
-            List<CompactionAggregate.UnifiedAggregate> aggregates = new ArrayList<>();
-            for (Bucket bucket : buckets)
-                aggregates.add(bucket.constructAggregate(controller, spaceAvailable, arena));
+            if (!buckets.isEmpty())
+            {
+                for (Bucket bucket : buckets)
+                    aggregates.add(bucket.constructAggregate(controller, spaceAvailable, arena));
+            }
+            else
+            {
+                // If there are no overlaps, we look if some shards have too many SSTables.
+                // If that's the case, we perform a major compaction on those shards.
```

Comment:
Nit: maybe add a "see CNDB-14577" as a further pointer to the reason we are doing these.

It also makes sense to extract this part in a method (I was surprised sonar isn't complaining about this, but it seems it isn't running).

Reply:
done